File: test_pages/login_page.py
"""Page Object Model for Login Page (/auth/login)."""

from selenium.webdriver.common.by import By
from test_pages.base_page import BasePage

class LoginPage(BasePage):
    # Locators with multi-strategy fallback
    EMAIL_INPUT = (By.XPATH, "//input[@name='email' or @type='email' or @placeholder='name@email.com']")
    PASSWORD_INPUT = (By.XPATH, "//input[@name='password' or @type='password' or @placeholder='••••••••']")
    SUBMIT_BUTTON = (By.XPATH, "//button[@type='submit' or contains(text(), 'Log In')]")
    FORGOT_PASSWORD_LINK = (By.XPATH, "//a[contains(@href, 'forgot-password') or contains(text(), 'Forgot')]")
    SIGNUP_LINK = (By.XPATH, "//a[contains(@href, 'signup') or contains(text(), 'Sign Up')]")
    ERROR_ALERT = (By.XPATH, "//div[contains(@className, 'rose') or contains(text(), 'Invalid') or contains(text(), 'error') or contains(text(), '⚠️')]")
    PASSWORD_TOGGLE = (By.XPATH, "//button[contains(@title, 'password') or .//*[name()='svg']]")

    def open(self):
        self.navigate_to("/auth/login")

    def enter_email(self, email):
        self.send_keys(*self.EMAIL_INPUT, email)

    def enter_password(self, password):
        self.send_keys(*self.PASSWORD_INPUT, password)

    def click_login(self):
        self.click(*self.SUBMIT_BUTTON)

    def login(self, email, password):
        self.enter_email(email)
        self.enter_password(password)
        self.click_login()

    def get_error_message(self):
        if self.is_visible(*self.ERROR_ALERT, timeout=3):
            return self.get_text(*self.ERROR_ALERT)
        return ""

File: selenium-node-tests/pages/LoginPage.ts
import { WebDriver } from 'selenium-webdriver';
import { BasePage } from './BasePage';

export class LoginPage extends BasePage {
  private readonly EMAIL = '#email, input[name="email"], input[type="email"]';
  private readonly PASSWORD = '#password, input[name="password"], input[type="password"]';
  private readonly SUBMIT = 'button[type="submit"], button#login-btn, [data-testid="login-button"]';
  private readonly ERROR = '.error-message, [role="alert"], [data-testid="error"]';
  private readonly LOADING = '.loading, [data-testid="loading"], button[disabled]';
  private readonly FORGOT = 'a[href*="forgot"], [data-testid="forgot-password"]';
  private readonly SIGN_UP = 'a[href*="signup"], a[href*="register"]';
  private readonly GOOGLE = '[data-testid="google-auth"], button.google-btn';

  constructor(driver: WebDriver) { super(driver); }

  async open(): Promise<void> { await this.navigate('/login'); }
  async enterEmail(email: string): Promise<void> { await this.type(this.EMAIL, email); }
  async enterPassword(password: string): Promise<void> { await this.type(this.PASSWORD, password); }
  async clickSubmit(): Promise<void> { await this.click(this.SUBMIT); await this.pause(800); }
  async clickForgotPassword(): Promise<void> { await this.click(this.FORGOT); }
  async clickSignUp(): Promise<void> { await this.click(this.SIGN_UP); }

  async login(email: string, password: string): Promise<void> {
    await this.enterEmail(email);
    await this.enterPassword(password);
    await this.clickSubmit();
  }

  async getErrorText(): Promise<string> {
    try { return this.getText(this.ERROR); } catch { return ''; }
  }
  async isErrorVisible(): Promise<boolean> { return this.isDisplayed(this.ERROR); }
  async isSubmitEnabled(): Promise<boolean> { return this.isEnabled(this.SUBMIT); }
  async waitForDashboard(): Promise<void> { await this.waitForURL('/dashboard'); }
}
